File: c++/11_Container_With_Most_Water.cpp
#include <iostream>
#include <vector>
using namespace std;


int maxArea(vector<int>& height) {
        int max_area = 0;
        int left = 0;
        int right = height.size() - 1;

        while (left < right){
            int area = (right - left) * min(height[left], height[right]);
            max_area = max(max_area, area);
            if (height[left] < height[right]){
                left++;
            }
            else{
                right--;

            }
        }

        return max_area;
    };


int main(){
    vector<int> heights = {1, 8, 6, 2, 5, 4, 8, 3, 7};
    cout << maxArea(heights) << endl;
};

File: py/11_Container_With_Most_Water.py
from typing import List


def maxArea(height: List[int]) -> int:
    max_area = 0
    left = 0
    right = len(height) - 1
    while left < right:
        area = min(height[left], height[right]) * (right - left)
        max_area = max(max_area, area)
        if height[left] < height[right]:
            left += 1
            continue  # continue 後 跳過當前迴圈迭代，進行下一次迭代
        right -= 1
    return max_area


if __name__ == "__main__":
    # unit test
    print(maxArea([1, 8, 6, 2, 5, 4, 8, 3, 7]))
